### PR TITLE
[FLINK-8005] [runtime] Set user code class loader before snapshot

### PR DESCRIPTION
**Pull request by GJL**
## What is the purpose of the change

*During checkpointing, user code may dynamically load classes from the user code
jar. This is a problem if the thread invoking the snapshot callbacks does not
have the user code class loader set as its context class loader. This commit
makes sure that the correct class loader is set.*

## Brief change log

  - *Set user code class loader in ThreadFactory of `Task#asyncCallDispatcher`*
  - *Clean up TaskAsyncCallTest*

## Verifying this change

This change added tests and can be verified as follows:
  - *Added unit tests to verify that context class loader is set*
  - *Started job with FlinkKafkaProducer011 and verified that snapshotting works*
 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java [hunk 1]
```diff
@@ -58,99 +58,119 @@
...
-			}
-			
-			task.cancelExecution();
-			task.getExecutingThread().join();
+		awaitLatch.await();
+
+		for (int i = 1; i <= numCalls; i++) {
+			task.triggerCheckpointBarrier(i, 156865867234L, CheckpointOptions.forCheckpoint());
 		}
-		catch (Exception e) {
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: Semantics of existing tests did not change: I removed the `try-catch` and simplified the assertion:

```
if (currentState != ExecutionState.RUNNING && currentState != ExecutionState.FINISHED) {
    fail("Task should be RUNNING or FINISHED, but is " + currentState);
}

```
to
```
assertThat(currentState, isOneOf(ExecutionState.RUNNING, ExecutionState.FINISHED));
```

**Comment by aljoscha** [comment]
These changes look good! 👍 

I'll wait for travis and then merge.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java (hunk 1 above)
Comment (aljoscha) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: Yep, the diff on GitHub is a bit hard to read but I figured it out. 😅

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskStopTest.java [hunk 2]
```diff
@@ -1,157 +0,0 @@
...
-		f.setAccessible(true);
-		f.set(task, taskMock);
-
-		Field f2 = task.getClass().getDeclaredField("executionState");
-		f2.setAccessible(true);
-		f2.set(task, ExecutionState.RUNNING);
-	}
-
-	@Test(timeout = 20000)
-	public void testStopExecution() throws Exception {
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskStopTest.java]: This only tested that stop is invoked. Should be covered by `testSetsUserCodeClassLoader`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java [hunk 3]
```diff
@@ -58,99 +59,144 @@
...
-			
+
 			assertFalse(task.isCanceledOrFailed());
 
 			ExecutionState currentState = task.getExecutionState();
-			if (currentState != ExecutionState.RUNNING && currentState != ExecutionState.FINISHED) {
-				fail("Task should be RUNNING or FINISHED, but is " + currentState);
-			}
-			
-			task.cancelExecution();
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: I moved this to the `AutoCloseable` `TaskCleaner`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java [hunk 4]
```diff
@@ -58,99 +59,144 @@
...
+
+			awaitLatch.await();
+
+			task.triggerCheckpointBarrier(1, 1, CheckpointOptions.forCheckpoint());
+			task.notifyCheckpointComplete(1);
+			task.stopExecution();
 		}
+
+		// assert after task is canceled and executing thread is stopped to avoid race conditions
+		assertThat(classLoaders, hasSize(greaterThanOrEqualTo(3)));
```
Comment (aljoscha) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: Are we guaranteed that all three calls have been made at this point or could this be flaky due to race conditions?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/DispatcherThreadFactory.java [hunk 5]
```diff
@@ -29,21 +31,41 @@
...
+
+	/**
+	 * Creates a new thread factory.
+	 *
+	 * @param group The group that the threads will be associated with.
+	 * @param threadName The name for the threads.
+	 * @param classLoader The {@link ClassLoader} to be set as context class loader.
+	 */
+	public DispatcherThreadFactory(
+		ThreadGroup group,
```
Comment (kl0u) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/DispatcherThreadFactory.java]: This is a code style preference rather than an issue, but I would suggest to indent the arguments by a tab to separate them from the body of the method.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java [hunk 6]
```diff
@@ -254,12 +300,10 @@ else if (this.error == null) {
 
 		@Override
 		public void triggerCheckpointOnBarrier(CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions, CheckpointMetrics checkpointMetrics) throws Exception {
-			throw new UnsupportedOperationException("Should not be called");
 		}
 
 		@Override
 		public void abortCheckpointOnBarrier(long checkpointId, Throwable cause) {
-			throw new UnsupportedOperationException("Should not be called");
```
Comment (kl0u) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: Why was this removed?

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java [hunk 7]
```diff
@@ -254,12 +300,10 @@ else if (this.error == null) {
 
 		@Override
 		public void triggerCheckpointOnBarrier(CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions, CheckpointMetrics checkpointMetrics) throws Exception {
-			throw new UnsupportedOperationException("Should not be called");
```
Comment (kl0u) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: Why was this removed?

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java (hunk 4 above)
Comment (kl0u) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: As @aljoscha suggested, I think that there is no guarantee that the 3 calls have finished by the time we check, right?

Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: I believe you are right. I introduced another latch to counter this.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/DispatcherThreadFactory.java (hunk 5 above)
Comment (GJL) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/DispatcherThreadFactory.java]: Indented.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java (hunk 7 above)
Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: Not sure but I decided to add it again.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java (hunk 6 above)
Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: Not sure anymore but I decided to add it again.

**Comment by GJL** [comment]
I addressed the comments. Let's wait for Travis and let me know if something else needs to be changed. 

@aljoscha  @kl0u

**Comment by aljoscha** [comment]
I think waiting on the stop latch might not be enough (in 100 % of cases) because the other two calls are also asynchronous.

**Comment by GJL** [comment]
There is only one thread dispatching the calls:
```
executor = Executors.newSingleThreadExecutor(
						new DispatcherThreadFactory(TASK_THREADS_GROUP, "Async calls on " + taskNameWithSubtask));
				this.asyncCallDispatcher = executor;
```

The tasks cannot overtake each other. I could make the test more strict and wait additionally on `triggerLatch` in case somebody decides to have multiple threads.

**Comment by aljoscha** [comment]
Yes, but I think this is making an assumption about the internal implementation. If someone changes that the test could break/not test the right thing anymore.

File: flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java [hunk 8]
```diff
@@ -277,8 +287,12 @@ public void invoke() throws Exception {
 				}
 			}
 
-			triggerLatch.trigger();
 			if (error != null) {
+				// exit method prematurely due to error but make sure that the tests can finish
+				triggerLatch.trigger();
```
Comment (kl0u) [inline flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: for all latches, it should also have:`if (!latch.isTriggered()) { latch.await() }`

Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: Why is that? I think at this point the latch might not get triggered at all.

Comment (kl0u) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: Sorry, I was just looking on the IDE and missed the lines. This line should be before every time you call `await` on the `latch`.

Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: I think it doesn't matter because the latch already checks for the flag.
```
	public void await() throws InterruptedException {
		synchronized (lock) {
			while (!triggered) {
				lock.wait();
			}
		}
	}
```

Comment (aljoscha) [reply flink-runtime/src/test/java/org/apache/flink/runtime/taskmanager/TaskAsyncCallTest.java]: yes, a latch that was already triggered will simply return immediately, no need for an additional check

**Comment by aljoscha** [comment]
thanks, I think this is excellent now. 👌 

I'll merge as soon as travis is green.

**Comment by kl0u** [comment]
I agree! +1 to merge as soon as Travis gives us the green light.

**Comment by aljoscha** [comment]
Thanks again for this fix! 👍 

Could you please close if GitHub doesn't auto-close?
